crates/agglayer-contracts/src/lib.rs

```diff
@@ -125,11 +154,109 @@ where
         self.default_l1_info_tree_entry
     }
 
-    async fn get_l1_info_root(&self, l1_leaf_count: u32) -> Result<[u8; 32], ()> {
-        self.l1_info_tree
-            .l_1_info_root_map(l1_leaf_count)
+    async fn get_l1_info_root(&self, l1_leaf_count: u32) -> Result<[u8; 32], L1RpcError> {
+        // Get `UpdateL1InfoTreeV2` event for the given leaf count from the latest block
+        let filter = Filter::new()
+            .address(self.l1_info_tree.address())
+            .event("UpdateL1InfoTreeV2(bytes32,uint32,uint256,uint64)")
+            .topic1(U256::from_big_endian(&l1_leaf_count.to_be_bytes()))
+            .from_block(BlockNumber::Earliest);
+
+        let events = self
+            .l1_info_tree
+            .client()
+            .get_logs(&filter)
             .await
-            .map_err(|_| ())
+            .map_err(|e| L1RpcError::UpdateL1InfoTreeV2EventFailure(e.to_string()))?;
+
+        // Extract event details
+        let (l1_info_root, event_block_number, event_block_hash) = events
+            .first()
+            .and_then(|log| {
+                match PolygonZkEVMGlobalExitRootV2Events::decode_log(&log.clone().into()).ok()? {
+                    PolygonZkEVMGlobalExitRootV2Events::UpdateL1InfoTreeV2Filter(event) => Some((
+                        event.current_l1_info_root,
+                        log.block_number?,
+                        log.block_hash?,
+                    )),
+                    _ => None,
+                }
+            })
+            .ok_or(L1RpcError::UpdateL1InfoTreeV2EventNotFound)?;
+
+        debug!(
+            "Retrieved UpdateL1InfoTreeV2 event from block {}. L1 info tree leaf count: {}, root: \
+             {}",
+            event_block_number,
+            l1_leaf_count,
+            H256::from_slice(l1_info_root.as_slice())
+        );
```

Comment:
Use structured logging.

Reply:
To do in another PR when we tackle this [one](https://github.com/agglayer/protocol-research/issues/59)